Doxygen/man/man3/Camera.3
.TH "Camera" 3 "Fri Nov 16 2012" "Version 001" "OpenGL Flythrough" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Camera \- 
.PP
The \fBCamera\fP class represents a logical camera in a model view, which posesses a current viewing angle and an absolute position in space as its state\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Camera\&.hpp>\fP
.SS "Public Types"

.in +1c
.ti -1c
.RI "enum \fBDirection\fP { \fBForward\fP, \fBBackward\fP, \fBLeft\fP, \fBRight\fP, \fBUp\fP, \fBDown\fP, \fBEnd\fP, \fBBegin\fP =  Forward }"
.br
.RI "\fIThe Direction enumeration lists all of the possible directions the camera may travel in\&. \fP"
.ti -1c
.RI "enum \fBglsl_var\fP { \fBTRANSLATION\fP, \fBROTATION\fP, \fBPERSPECTIVE\fP, \fBTRP_M\fP, \fBPRT_M\fP, \fBCTM\fP, \fBNumGlslVars\fP }"
.br
.RI "\fIThe glsl_var enumeration lists the various variables the \fBCamera\fP class is capable of sending to the shader\&. \fP"
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBCamera\fP (float x=0\&.0, float y=0\&.0, float z=0\&.0)"
.br
.RI "\fIInitialization Constructor; sets the X,Y,Z coordinates explicitly\&. \fP"
.ti -1c
.RI "\fBCamera\fP (vec3 &in)"
.br
.RI "\fIInitialization Constructor, uses a vec3 as its initial coordinates\&. \fP"
.ti -1c
.RI "\fBCamera\fP (vec4 &in)"
.br
.RI "\fIInitialization Constructor, uses a vec4 as its initial coordinates\&. \fP"
.ti -1c
.RI "virtual \fB~Camera\fP (void)"
.br
.RI "\fIDefault destructor\&. \fP"
.ti -1c
.RI "void \fBX\fP (const float &in, const bool &update=true)"
.br
.RI "\fISets the X coordinate of the camera\&. \fP"
.ti -1c
.RI "void \fBY\fP (const float &in, const bool &update=true)"
.br
.RI "\fISets the Y coordinate of the camera\&. \fP"
.ti -1c
.RI "void \fBZ\fP (const float &in, const bool &update=true)"
.br
.RI "\fISets the Z coordinate of the camera\&. \fP"
.ti -1c
.RI "void \fBpos\fP (const float &x, const float &y, const float &z, const bool &update=true)"
.br
.RI "\fISets the absolute position of the camera\&. \fP"
.ti -1c
.RI "void \fBpos\fP (const vec3 &in, const bool &update=true)"
.br
.RI "\fISets the absolute position of the camera\&. \fP"
.ti -1c
.RI "void \fBpos\fP (const vec4 &in, const bool &update=true)"
.br
.RI "\fISets the absolute position of the camera\&. \fP"
.ti -1c
.RI "void \fBdX\fP (const float &by, const bool &update=true)"
.br
.RI "\fIMoves the camera along the X axis\&. \fP"
.ti -1c
.RI "void \fBdY\fP (const float &by, const bool &update=true)"
.br
.RI "\fIMoves the camera along the Y axis\&. \fP"
.ti -1c
.RI "void \fBdZ\fP (const float &by, const bool &update=true)"
.br
.RI "\fIMoves the camera along the Z axis\&. \fP"
.ti -1c
.RI "void \fBdPos\fP (const float &x, const float &y, const float &z)"
.br
.RI "\fIMoves the camera along the x, y, and z axes\&. \fP"
.ti -1c
.RI "void \fBdPos\fP (const vec3 &by)"
.br
.RI "\fIMoves the camera along the x, y, and z axes\&. \fP"
.ti -1c
.RI "void \fBdPos\fP (const vec4 &by)"
.br
.RI "\fIMoves the camera along the x, y, and z axes\&. \fP"
.ti -1c
.RI "void \fBFOV\fP (const float &\fBfovy\fP)"
.br
.RI "\fIFOV sets the current camera Field-of-view angle\&. \fP"
.ti -1c
.RI "float \fBFOV\fP (void) const "
.br
.RI "\fI\fBFOV()\fP gets the current camera Field-of-view angle\&. \fP"
.ti -1c
.RI "void \fBdFOV\fP (const float &by)"
.br
.RI "\fIdFOV adjusts the field of view angle up or down by an amount\&. \fP"
.ti -1c
.RI "void \fBsway\fP (const float &by)"
.br
.RI "\fIAdjusts the camera's X coordinate relative to its current position\&. \fP"
.ti -1c
.RI "void \fBsurge\fP (const float &by)"
.br
.RI "\fIAdjusts the camera's Z coordinate relative to its current position\&. \fP"
.ti -1c
.RI "void \fBheave\fP (const float &by)"
.br
.RI "\fIAdjusts the camera's Y coordinate relative to its current position\&. \fP"
.ti -1c
.RI "void \fBpitch\fP (const float &by)"
.br
.RI "\fIpitch adjusts the X axis rotation; up/down look\&. \fP"
.ti -1c
.RI "void \fByaw\fP (const float &by)"
.br
.RI "\fIyaw adjusts the Y axis rotation; left/right look\&. \fP"
.ti -1c
.RI "void \fBroll\fP (const float &by)"
.br
.RI "\fIroll adjusts the Z axis rotation; tilt or lean left/right\&. \fP"
.ti -1c
.RI "void \fBMove\fP (const \fBCamera::Direction\fP &Dir)"
.br
.RI "\fIMove instructs the camera to begin moving in the specified direction\&. \fP"
.ti -1c
.RI "void \fBStop\fP (const \fBCamera::Direction\fP &Dir)"
.br
.RI "\fIStop instructs the camera to stop moving in the specified direction\&. \fP"
.ti -1c
.RI "void \fBIdle\fP (void)"
.br
.RI "\fIIdle moves the camera forward in whichever directions it is configured to move in\&. \fP"
.ti -1c
.RI "float \fBX\fP (void) const "
.br
.RI "\fI\fBX()\fP returns the current position of the camera in model coordinates\&. \fP"
.ti -1c
.RI "float \fBY\fP (void) const "
.br
.RI "\fI\fBY()\fP returns the current position of the camera in model coordinates\&. \fP"
.ti -1c
.RI "float \fBZ\fP (void) const "
.br
.RI "\fI\fBZ()\fP returns the current position of the camera in model coordinates\&. \fP"
.ti -1c
.RI "vec4 \fBpos\fP (void) const "
.br
.RI "\fI\fBpos()\fP gets the current camera position in model coordinates\&. \fP"
.ti -1c
.RI "void \fBsend\fP (const \fBglsl_var\fP &which)"
.br
.RI "\fIsend will send a glsl variable to the shader\&. \fP"
.ti -1c
.RI "void \fBlink\fP (GLuint &program, const \fBglsl_var\fP &which, const string &glslVarName)"
.br
.RI "\fILink associates the camera with a glsl uniform variable\&. \fP"
.in -1c
.SS "Static Public Attributes"

.in +1c
.ti -1c
.RI "static const float \fBSpeed\fP = 0\&.01"
.br
.RI "\fIThe increment the camera should move during the \fBIdle()\fP function\&. \fP"
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "void \fBadjustRotation\fP (const mat4 &adjustment)"
.br
.RI "\fIadjustRotation is an internal function that rotates the camera\&. \fP"
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "mat4 \fBT\fP"
.br
.ti -1c
.RI "mat4 \fBR\fP"
.br
.ti -1c
.RI "mat4 \fBP\fP"
.br
.ti -1c
.RI "mat4 \fBTRP\fP"
.br
.ti -1c
.RI "mat4 \fBPRT\fP"
.br
.ti -1c
.RI "mat4 \fBctm\fP"
.br
.ti -1c
.RI "GLfloat \fBfovy\fP"
.br
.RI "\fICurrent field-of-view angle for perspective view\&. \fP"
.ti -1c
.RI "bool \fBMotion\fP [Camera::End]"
.br
.RI "\fIBooleans correlating to the different motion directions\&. \fP"
.ti -1c
.RI "GLuint \fBglsl_handles\fP [Camera::NumGlslVars]"
.br
.RI "\fIHandles for communicating with the shader\&. \fP"
.in -1c
.SH "Detailed Description"
.PP 
The \fBCamera\fP class represents a logical camera in a model view, which posesses a current viewing angle and an absolute position in space as its state\&. 

\fBAuthor:\fP
.RS 4
John Huston, jhuston@cs.uml.edu 
.RE
.PP
\fBSince:\fP
.RS 4
16 Nov 2012
.RE
.PP
Functions are provided to adjust the rotation according to \fBpitch()\fP, \fByaw()\fP and \fBroll()\fP motions; \fBsurge()\fP, \fBsway()\fP, and \fBheave()\fP are provided to adjust position in space\&.
.PP
\fBMove()\fP, \fBStop()\fP, and \fBIdle()\fP are provided to help the camera automatically move along the X, Y, or Z axes\&. 
.PP
Definition at line 23 of file Camera\&.hpp\&.
.SH "Member Enumeration Documentation"
.PP 
.SS "enum \fBCamera::Direction\fP"
.PP
The Direction enumeration lists all of the possible directions the camera may travel in\&. 'Begin' and 'End' are special sentinel directions for the purposes of iteration, and are ignored by any functions that accept a Direction\&. 
.PP
Definition at line 33 of file Camera\&.hpp\&.
.SS "enum \fBCamera::glsl_var\fP"
.PP
The glsl_var enumeration lists the various variables the \fBCamera\fP class is capable of sending to the shader\&. The NumGlslVars variable is a sentinel value that is ignored by any functions that accept a glsl_var\&. 
.PP
Definition at line 51 of file Camera\&.hpp\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "\fBCamera::Camera\fP (floatx = \fC0\&.0\fP, floaty = \fC0\&.0\fP, floatz = \fC0\&.0\fP)"
.PP
Initialization Constructor; sets the X,Y,Z coordinates explicitly\&. \fBParameters:\fP
.RS 4
\fIx\fP The initial X coordinate\&. 
.br
\fIy\fP The initial Y coordinate\&. 
.br
\fIz\fP The initial Z coordinate\&. 
.RE
.PP

.PP
Definition at line 14 of file Camera\&.cpp\&.
.SS "\fBCamera::Camera\fP (vec3 &in)"
.PP
Initialization Constructor, uses a vec3 as its initial coordinates\&. \fBParameters:\fP
.RS 4
\fIin\fP A vec3 representing the initial coordinates\&. 
.RE
.PP

.PP
Definition at line 24 of file Camera\&.cpp\&.
.SS "\fBCamera::Camera\fP (vec4 &in)"
.PP
Initialization Constructor, uses a vec4 as its initial coordinates\&. \fBParameters:\fP
.RS 4
\fIin\fP A vec4 representing the initial coordinates\&. The w component is ignored\&. 
.RE
.PP

.PP
Definition at line 33 of file Camera\&.cpp\&.
.SS "\fBCamera::~Camera\fP (void)\fC [virtual]\fP"
.PP
Default destructor\&. Nothing of note\&. 
.PP
Definition at line 41 of file Camera\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "void \fBCamera::adjustRotation\fP (const mat4 &adjustment)\fC [private]\fP"
.PP
adjustRotation is an internal function that rotates the camera\&. Technically, any transformation, not just a rotation, is possible\&. 
.PP
\fBParameters:\fP
.RS 4
\fIadjustment\fP The 4x4 matrix to transform the CTM by\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
Void\&. 
.RE
.PP

.PP
Definition at line 197 of file Camera\&.cpp\&.
.SS "void \fBCamera::dFOV\fP (const float &by)"
.PP
dFOV adjusts the field of view angle up or down by an amount\&. \fBParameters:\fP
.RS 4
\fIby\fP The float to adjust the FOV angle by\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
Void\&. 
.RE
.PP

.PP
Definition at line 405 of file Camera\&.cpp\&.
.SS "void \fBCamera::dPos\fP (const float &x, const float &y, const float &z)"
.PP
Moves the camera along the x, y, and z axes\&. \fBParameters:\fP
.RS 4
\fIx\fP the X-axis displacement\&. 
.br
\fIy\fP the Y-axis displacement\&. 
.br
\fIz\fP the Z-axis displacement\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
Void\&. 
.RE
.PP

.PP
Definition at line 162 of file Camera\&.cpp\&.
.SS "void \fBCamera::dPos\fP (const vec3 &by)"
.PP
Moves the camera along the x, y, and z axes\&. \fBParameters:\fP
.RS 4
\fIby\fP A vec3 containing the X, Y, and Z axis displacements\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
Void\&. 
.RE
.PP

.PP
Definition at line 176 of file Camera\&.cpp\&.
.SS "void \fBCamera::dPos\fP (const vec4 &by)"
.PP
Moves the camera along the x, y, and z axes\&. \fBParameters:\fP
.RS 4
\fIby\fP A vec4 containing the X, Y, and Z axis displacements\&. The w component is ignored\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
Void\&. 
.RE
.PP

.PP
Definition at line 186 of file Camera\&.cpp\&.
.SS "void \fBCamera::dX\fP (const float &by, const bool &update = \fCtrue\fP)"
.PP
Moves the camera along the X axis\&. \fBParameters:\fP
.RS 4
\fIby\fP The float value of the X-axis displacement\&. 
.br
\fIupdate\fP A boolean indicating whether or not to update the shader\&. update defaults to true\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
void\&. 
.RE
.PP

.PP
Definition at line 126 of file Camera\&.cpp\&.
.SS "void \fBCamera::dY\fP (const float &by, const bool &update = \fCtrue\fP)"
.PP
Moves the camera along the Y axis\&. \fBParameters:\fP
.RS 4
\fIby\fP The float value of the Y-axis displacement\&. 
.br
\fIupdate\fP A boolean indicating whether or not to update the shader\&. update defaults to true\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
Void\&. 
.RE
.PP

.PP
Definition at line 138 of file Camera\&.cpp\&.
.SS "void \fBCamera::dZ\fP (const float &by, const bool &update = \fCtrue\fP)"
.PP
Moves the camera along the Z axis\&. \fBParameters:\fP
.RS 4
\fIby\fP The float value of the Z-axis displacement\&. 
.br
\fIupdate\fP A boolean indicating whether or not to update the shader\&. update defaults to true\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
Void\&. 
.RE
.PP

.PP
Definition at line 150 of file Camera\&.cpp\&.
.SS "void \fBCamera::FOV\fP (const float &in)"
.PP
FOV sets the current camera Field-of-view angle\&. This function will send the new perspective matrix to the shader\&. 
.PP
\fBParameters:\fP
.RS 4
\fIin\fP The new field of view angle\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
Void\&. 
.RE
.PP

.PP
Definition at line 389 of file Camera\&.cpp\&.
.SS "float \fBCamera::FOV\fP (void) const"
.PP
\fBFOV()\fP gets the current camera Field-of-view angle\&. \fBReturns:\fP
.RS 4
A float that is the y axis viewing angle\&. 
.RE
.PP

.PP
Definition at line 380 of file Camera\&.cpp\&.
.SS "void \fBCamera::heave\fP (const float &by)"
.PP
Adjusts the camera's Y coordinate relative to its current position\&. Positive values move the camera up, and negative values move the camera down\&. 
.PP
\fBParameters:\fP
.RS 4
\fIby\fP The float to adjust the Y coordinate by\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
Void\&. 
.RE
.PP

.PP
Definition at line 254 of file Camera\&.cpp\&.
.SS "void \fBCamera::Idle\fP (void)"
.PP
Idle moves the camera forward in whichever directions it is configured to move in\&. Call it in the glut Idle function\&. 
.PP
\fBReturns:\fP
.RS 4
Void\&. 
.RE
.PP

.PP
Definition at line 334 of file Camera\&.cpp\&.
.SS "void \fBCamera::link\fP (GLuint &program, const \fBglsl_var\fP &which, const string &glslVarName)"
.PP
Link associates the camera with a glsl uniform variable\&. \fBParameters:\fP
.RS 4
\fIprogram\fP a GLuint handle to the shader application\&. 
.br
\fIwhich\fP A glsl_var enumeration indication which variable to link\&. 
.br
\fIglslVarName\fP The name of the variable in the shader\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
Void\&. 
.RE
.PP

.PP
Definition at line 458 of file Camera\&.cpp\&.
.SS "void \fBCamera::Move\fP (const \fBCamera::Direction\fP &Dir)"
.PP
Move instructs the camera to begin moving in the specified direction\&. \fBParameters:\fP
.RS 4
\fIDir\fP The direction in which to move\&. Can be any direction in the enumerated type \fBCamera::Direction\fP\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
Void\&. 
.RE
.PP

.PP
Definition at line 314 of file Camera\&.cpp\&.
.SS "void \fBCamera::pitch\fP (const float &by)"
.PP
pitch adjusts the X axis rotation; up/down look\&. A positive value represents looking up, while a negative value represents looking down\&. 
.PP
\fBParameters:\fP
.RS 4
\fIby\fP A float, in degrees, to adjust the pitch by\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
Void\&. 
.RE
.PP

.PP
Definition at line 266 of file Camera\&.cpp\&.
.SS "void \fBCamera::pos\fP (const float &x, const float &y, const float &z, const bool &update = \fCtrue\fP)"
.PP
Sets the absolute position of the camera\&. \fBParameters:\fP
.RS 4
\fIx\fP The new X coordinate of the camera\&. 
.br
\fIy\fP The new Y coordinate of the camera\&. 
.br
\fIz\fP The new Z coordinate of the camera\&. 
.br
\fIupdate\fP Whether or not to update the shader with the new coordinates\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
Void\&. 
.RE
.PP

.PP
Definition at line 88 of file Camera\&.cpp\&.
.SS "void \fBCamera::pos\fP (const vec3 &in, const bool &update = \fCtrue\fP)"
.PP
Sets the absolute position of the camera\&. \fBParameters:\fP
.RS 4
\fIin\fP A vec3 containing the x, y, and z coordinates to set the camera to\&. 
.br
\fIupdate\fP Whether or not to update the shader with the new coordinates\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
Void\&. 
.RE
.PP

.PP
Definition at line 114 of file Camera\&.cpp\&.
.SS "void \fBCamera::pos\fP (const vec4 &in, const bool &update = \fCtrue\fP)"
.PP
Sets the absolute position of the camera\&. \fBParameters:\fP
.RS 4
\fIin\fP A vec4 containing the x, y, and z coordinates to set the camera to\&. The w coordinate is ignored\&. 
.br
\fIupdate\fP Whether or not to update the shader with the new coordinates\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
Void\&. 
.RE
.PP

.PP
Definition at line 103 of file Camera\&.cpp\&.
.SS "vec4 \fBCamera::pos\fP (void) const"
.PP
\fBpos()\fP gets the current camera position in model coordinates\&. \fBReturns:\fP
.RS 4
A vec4 that represents the current camera coordinates\&. 
.RE
.PP

.PP
Definition at line 373 of file Camera\&.cpp\&.
.SS "void \fBCamera::roll\fP (const float &by)"
.PP
roll adjusts the Z axis rotation; tilt or lean left/right\&. A positive value represents leaning right, while a negative value represents leaning left\&. 
.PP
\fBParameters:\fP
.RS 4
\fIby\fP A float, in degrees, to adjust the roll by\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
Void\&. 
.RE
.PP

.PP
Definition at line 303 of file Camera\&.cpp\&.
.SS "void \fBCamera::send\fP (const \fBglsl_var\fP &which)"
.PP
send will send a glsl variable to the shader\&. \fBParameters:\fP
.RS 4
\fIwhich\fP The parameter to send\&. Can be any from enum glsl_var\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
Void\&. 
.RE
.PP

.PP
Definition at line 415 of file Camera\&.cpp\&.
.SS "void \fBCamera::Stop\fP (const \fBCamera::Direction\fP &Dir)"
.PP
Stop instructs the camera to stop moving in the specified direction\&. \fBParameters:\fP
.RS 4
\fIDir\fP The direction in which to stop moving\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
Void\&. 
.RE
.PP

.PP
Definition at line 324 of file Camera\&.cpp\&.
.SS "void \fBCamera::surge\fP (const float &by)"
.PP
Adjusts the camera's Z coordinate relative to its current position\&. Positive values move the camera forward, and negative values move the camera backward\&. Note that the camera uses model coordinates internally, so moving forward will increase the camera's Z position negatively\&. 
.PP
\fBParameters:\fP
.RS 4
\fIby\fP The float to adjust the Z coordinate by\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
Void\&. 
.RE
.PP

.PP
Definition at line 242 of file Camera\&.cpp\&.
.SS "void \fBCamera::sway\fP (const float &by)"
.PP
Adjusts the camera's X coordinate relative to its current position\&. Negative values move the camera left, and positive values move the camera right\&. 
.PP
\fBParameters:\fP
.RS 4
\fIby\fP The float to adjust the X coordinate by\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
Void\&. 
.RE
.PP

.PP
Definition at line 228 of file Camera\&.cpp\&.
.SS "void \fBCamera::X\fP (const float &in, const bool &update = \fCtrue\fP)"
.PP
Sets the X coordinate of the camera\&. \fBParameters:\fP

Flythrough/Camera.cpp
#include <stdexcept>
#include "Angel.h"
#include "Camera.hpp"


const float Camera::Speed = 0.01;

/**
   Initialization Constructor; sets the X,Y,Z coordinates explicitly.
   @param x The initial X coordinate.
   @param y The initial Y coordinate.
   @param z The initial Z coordinate.
**/
Camera::Camera( float x, float y, 
		float z ) {
  this->pos( x, y, z, false );
}


/**
   Initialization Constructor, uses a vec3 as its initial coordinates.
   @param in A vec3 representing the initial coordinates.
**/
Camera::Camera( vec3 &in ) {
  this->pos( in, false );
}


/**
   Initialization Constructor, uses a vec4 as its initial coordinates.
   @param in A vec4 representing the initial coordinates. The w component is ignored.
**/
Camera::Camera( vec4 &in ) {
  this->pos( in, false );
}


/**
   Default destructor. Nothing of note.
**/
Camera::~Camera( void ) { }


/**
   Sets the X coordinate of the camera.
   @param in     The new X coordinate of the camera.
   @param update Whether or not to update the shader with the new coordinates.
   @return Void.
**/
void Camera::X( const float &in, const bool &update ) { 
  T[0][3] = -in;
  if (update) send( TRANSLATION );
}


/**
   Sets the Y coordinate of the camera.
   @param in     The new Y coordinate of the camera.
   @param update Whether or not to update the shader with the new coordinates.
   @return Void.
**/
void Camera::Y( const float &in, const bool &update ) { 
  T[1][3] = -in;
  if (update) send( TRANSLATION );
}


/**
   Sets the Z coordinate of the camera.
   @param in     The new Z coordinate of the camera.
   @param update Whether or not to update the shader with the new coordinates.
   @return Void.
**/
void Camera::Z( const float &in, const bool &update ) { 
  T[2][3] = -in;
  if (update) send( TRANSLATION );
}


/**
   Sets the absolute position of the camera.
   @param x The new X coordinate of the camera.
   @param y The new Y coordinate of the camera.
   @param z The new Z coordinate of the camera.
   @param update Whether or not to update the shader with the new coordinates.
   @return Void.
**/
void Camera::pos( const float &x, const float &y, 
		  const float &z, const bool &update ) {
  X(x, false);
  Y(y, false);
  Z(z, false);
  if (update) send( TRANSLATION );
}


/**
   Sets the absolute position of the camera.
   @param in A vec4 containing the x, y, and z coordinates to set the camera to. The w coordinate is ignored.
   @param update Whether or not to update the shader with the new coordinates.
   @return Void.
**/
void Camera::pos( const vec4 &in, const bool &update ) {
  this->pos( in.x, in.y, in.z, update );
}


/**
   Sets the absolute position of the camera.
   @param in A vec3 containing the x, y, and z coordinates to set the camera to.
   @param update Whether or not to update the shader with the new coordinates.
   @return Void.
**/
void Camera::pos( const vec3 &in, const bool &update ) {
  this->pos( in.x, in.y, in.z, update );
}

  
/**
   Moves the camera along the X axis.
   @param by The float value of the X-axis displacement.
   @param update A boolean indicating whether or not to update the shader. 
   update defaults to true.
   @return void.
**/
void Camera::dX( const float &by, const bool &update ) { 
  X( X() + by, update ); 
}


/**
   Moves the camera along the Y axis.
   @param by The float value of the Y-axis displacement.
   @param update A boolean indicating whether or not to update the shader.
   update defaults to true.
   @return Void.
**/
void Camera::dY( const float &by, const bool &update ) {
  Y( Y() + by, update );
}


/**
   Moves the camera along the Z axis.
   @param by The float value of the Z-axis displacement.
   @param update A boolean indicating whether or not to update the shader.
   update defaults to true.
   @return Void.
**/
void Camera::dZ( const float &by, const bool &update ) { 
  Z( Z() + by, update ); 
}


/** 
    Moves the camera along the x, y, and z axes.
    @param x the X-axis displacement.
    @param y the Y-axis displacement.
    @param z the Z-axis displacement.
    @return Void.
**/
void Camera::dPos( const float &x, const float &y, 
		   const float &z ) { 
  dX( x, false );
  dY( y, false );
  dZ( z, false );
  send( TRANSLATION );
}


/** 
    Moves the camera along the x, y, and z axes.
    @param by A vec3 containing the X, Y, and Z axis displacements.
    @return Void.
**/
void Camera::dPos( const vec3 &by ) {
  dPos( by.x, by.y, by.z );
}


/** 
    Moves the camera along the x, y, and z axes.
    @param by A vec4 containing the X, Y, and Z axis displacements. The w component is ignored.
    @return Void.
**/
void Camera::dPos( const vec4 &by ) {
  dPos( by.x, by.y, by.z );
}


/** 
    adjustRotation is an internal function that rotates the camera.
    Technically, any transformation, not just a rotation, is possible.
    @param adjustment The 4x4 matrix to transform the CTM by.
    @return Void.
**/
void Camera::adjustRotation( const mat4 &adjustment ) {
#ifdef POSTMULT
  // In a post-mult system, the argument order is left-to-right,
  // So the adjustment appears last.
  R = R * adjustment;
#else
  // In a pre-mult system, the last argument is applied first,
  // So the adjustment should appear first.
  R = adjustment * R;
#endif
  send( ROTATION );
}

/** 
  ROTATE_OFFSET is a macro which is used to normalize
  the six camera motion directions with respect to the
  current camera rotation. It is used in heave(), sway() and surge().
  @param V a vec4 representing the movement offset vector.
  @return A rotated vec4.
**/
//#define ROTATE_OFFSET(V) (transpose(R) * V)
#define ROTATE_OFFSET(V) (V * R)


/**
   Adjusts the camera's X coordinate relative to its current position.
   Negative values move the camera left,
   and positive values move the camera right.
   @param by The float to adjust the X coordinate by.
   @return Void.
**/
void Camera::sway( const float &by ) {
  dPos(ROTATE_OFFSET(vec4(by,0,0,0)));
}


/**
   Adjusts the camera's Z coordinate relative to its current position.
   Positive values move the camera forward,
   and negative values move the camera backward.
   Note that the camera uses model coordinates internally,
   so moving forward will increase the camera's Z position negatively.
   @param by The float to adjust the Z coordinate by.
   @return Void.
**/
void Camera::surge( const float &by ) {
  dPos(ROTATE_OFFSET(vec4(0,0,-by,0)));
}


/**
   Adjusts the camera's Y coordinate relative to its current position.
   Positive values move the camera up,
   and negative values move the camera down.
   @param by The float to adjust the Y coordinate by.
   @return Void.
**/
void Camera::heave( const float &by ) {
  dPos(ROTATE_OFFSET(vec4(0,by,0,0)));
}


/**
   pitch adjusts the X axis rotation; up/down look.
   A positive value represents looking up,
   while a negative value represents looking down.
   @param by A float, in degrees, to adjust the pitch by.
   @return Void.
**/
void Camera::pitch( const float &by ) {
  /*
    Since negative values are interpreted as pitching down,
    We leave the input uninverted, because a negative rotation
    about the X axis rotates counter-clockwise (looking right),
    and clockwise (looking left), which achieves the effect of
    looking 'down'.
  */ 
  adjustRotation(RotateX(-by));
}


/**
   yaw adjusts the Y axis rotation; left/right look.
   A positive value represents looking right,
   while a negative value represents looking left.
   @param by A float, in degrees, to adjust the yaw by.
   @return Void.
**/
void Camera::yaw( const float &by ) {
  /*
    Since a positive 'by' should represent looking right,
    we invert the rotation because rotating by a positive value
    will rotate right, which simulates looking left.
    Therefore, invert.
  */
  adjustRotation(RotateY(by));
}


/**
   roll adjusts the Z axis rotation; tilt or lean left/right.
   A positive value represents leaning right,
   while a negative value represents leaning left.
   @param by A float, in degrees, to adjust the roll by.
   @return Void.
**/
void Camera::roll( const float &by ) {
  adjustRotation(RotateZ(by));
}


/**
   Move instructs the camera to begin moving in the specified direction. 
   @param Dir The direction in which to move.
     Can be any direction in the enumerated type Camera::Direction.
   @return Void.
**/
void Camera::Move( const Camera::Direction &Dir ) { 
  Motion[Dir] = true; 
}


/**
   Stop instructs the camera to stop moving in the specified direction.
   @param Dir The direction in which to stop moving.
   @return Void.
**/
void Camera::Stop( const Camera::Direction &Dir ) {
  Motion[Dir] = false;
}


/**
   Idle moves the camera forward in whichever directions it is
   configured to move in. Call it in the glut Idle function.
   @return Void.
**/
void Camera::Idle( void ) {

  if (Motion[Camera::Forward]) surge( Camera::Speed );
  else if (Motion[Camera::Backward]) surge( -Camera::Speed );

  if (Motion[Camera::Right]) sway( Camera::Speed );
  else if (Motion[Camera::Left]) sway( -Camera::Speed );

  if (Motion[Camera::Up]) heave( Camera::Speed );
  else if (Motion[Camera::Down]) heave( -Camera::Speed );

}


/**
   X() returns the current position of the camera in model coordinates.
   @return The current X coordinate of the camera in model coordinates.
**/
float Camera::X( void ) const { return -T[0][3]; }


/**
   Y() returns the current position of the camera in model coordinates.
   @return The current Y coordinate of the camera in model coordinates.
**/
float Camera::Y( void ) const { return -T[1][3]; }


/**
   Z() returns the current position of the camera in model coordinates.
   @return The current Z coordinate of the camera in model coordinates.
**/
float Camera::Z( void ) const { return -T[2][3]; }


/**
   pos() gets the current camera position in model coordinates.
   @return A vec4 that represents the current camera coordinates.
**/
vec4 Camera::pos( void ) const { return vec4( X(), Y(), Z(), 1.0 ); }


/**
   FOV() gets the current camera Field-of-view angle.
   @return A float that is the y axis viewing angle.
**/
float Camera::FOV( void ) const { return fovy; }


/**
   FOV sets the current camera Field-of-view angle.
   This function will send the new perspective matrix to the shader.
   @param in The new field of view angle.
   @return Void.
**/
void Camera::FOV( const float &in ) { 
  GLint size[4];
  fovy = in;
  glGetIntegerv( GL_VIEWPORT, size );  
  P = Perspective( fovy,
		   (float)size[2]/(float)size[3],
		   0.3, 3.0 );
  send( PERSPECTIVE );
}


/**
   dFOV adjusts the field of view angle up or down by an amount.
   @param by The float to adjust the FOV angle by.
   @return Void.
**/
void Camera::dFOV( const float &by ) {
  FOV( FOV() + by );
}


/**
   send will send a glsl variable to the shader.
   @param which The parameter to send. Can be any from enum glsl_var.
   @return Void.
**/
void Camera::send( const glsl_var &which ) {

  switch (which) {
  case TRANSLATION:
    glUniformMatrix4fv( glsl_handles[which], 1, GL_FALSE, T );
    send( PRT_M );
    break;
  case ROTATION:
    glUniformMatrix4fv( glsl_handles[which], 1, GL_FALSE, R );
    send( PRT_M );
    break;
  case PERSPECTIVE:
    glUniformMatrix4fv( glsl_handles[which], 1, GL_FALSE, P );
    send( PRT_M );
    break;
  case TRP_M:
    TRP = T*R*P;
    glUniformMatrix4fv( glsl_handles[which], 1, GL_FALSE, TRP );
    break;
  case PRT_M:
    PRT = P*R*T;
    glUniformMatrix4fv( glsl_handles[which], 1, GL_TRUE, PRT );
    break;
  case CTM:
    glUniformMatrix4fv( glsl_handles[which], 1, GL_FALSE, ctm );
    break;
  default:
    throw std::invalid_argument( "Unknown GLSL variable handle." );
  }
}


/**
   Link associates the camera with a glsl uniform variable.
   @param program a GLuint handle to the shader application.
   @param which A glsl_var enumeration indication which variable to link.
   @param glslVarName The name of the variable in the shader.
   @return Void.
**/
void Camera::link( GLuint &program, const glsl_var &which, 
		   const string &glslVarName ) {
  
  glsl_handles[which] = glGetUniformLocation( program, 
					      glslVarName.c_str() );
  send( which );

}
